tests/conftest.py

```diff
@@ -21,6 +24,34 @@ def _inner(file_contents: Dict[str, str]) -> Path:
     return _inner
 
 
+@pytest.fixture
+def fake_venv(tmp_path):
+    def create_one_fake_venv(fake_packages: Dict[str, Set[str]]) -> Path:
+        venv_dir = Path(mkdtemp(prefix="fake_venv.", dir=tmp_path))
+        venv.create(venv_dir, with_pip=False)
+
+        # Create fake packages
+        major, minor = sys.version_info[:2]
+        site_dir = venv_dir / f"lib/python{major}.{minor}/site-packages"
+        assert site_dir.is_dir()
+        for package_name, import_names in fake_packages.items():
+            # Create just enough files under site_dir to fool importlib_metadata
+            # into believing these are genuine packages
+            dist_info_dir = site_dir / f"{package_name}-1.2.3.dist-info"
+            dist_info_dir.mkdir()
+            (dist_info_dir / "METADATA").write_text(
+                f"Name: {package_name}\nVersion: 1.2.3\n"
+            )
+            top_level = dist_info_dir / "top_level.txt"
+            top_level.write_text("".join(f"{name}\n" for name in sorted(import_names)))
+            for name in import_names:
+                (site_dir / f"{name}.py").touch()
```

Comment:
So, I'm not sure if there's an "official" definition of what constitutes a minimal Python package (and looking at wheels vs. eggs vs. source archives would certainly complicate that definition).

However after digging around inside `importlib_metadata` for a few days now, _this_ is what I found to be the minimal needed, specifically for `importlib_metadata`, yes:
- The `.dist-info` subdir is how wheels are installed, so I chose that as the more relevant thing to fake (as opposed to the deprecated-but-still-in-use `.egg-info` format).
- The `METADATA` files has a _lot_ of fields for most packages, but the `Name` is what `importlib_metadata` uses to identify the package name.
- I'm actually not strictly sure that the `Version` is necessary, but it's also part of the `*.dist-info` subdir name, so I thought it best to not rock the boat too much.
- The `top_level.txt` is the _preferred_ way to enumerate what import names a package provides.
- The `.py` files are dummy modules, but in real life they are most often _packages_ instead, i.e. subdirs with `__init__.py` files and more. _Technically_, with the current version of `importlib_metadata` I don't think these are necessary, since the `top_level.txt` is followed blindly. But my upstream PR at https://github.com/python/importlib_metadata/pull/437 might end up checking that these packages/modules actually exist (to work around bugs in `_top_level_inferred()`).

Reply:
Thanks. This makes sense now.